feat: Custom Action Sample

luhan2017 opened the pull request:
## Description

<!---
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context.

If this is a bug fix, please describe the root cause and analysis of this problem.
---->

In this PR, I added a CustomAction solution in the C# runtime folder which includes the custom code and schema.

Here are the steps to enable CustomAction:
1. Create a echo bot. 
2. Enable ejected runtime, the runtime will be placed in the bot folder.
3. In the bot/runtime folder add a project reference in azurewebapp/Microsoft.BotFramework.Composer.WebApp.csproj, 
```
<ProjectReference Include="..\customaction\Microsoft.BotFramework.Composer.CustomAction.csproj" />
```
4. Then remove line 28, uncomment line 29 and 139 in azurewebapp/Startup.cs file
```
using Microsoft.BotFramework.Composer.CustomAction;
// This is for custom action component registration.
ComponentRegistration.Add(new CustomActionComponentRegistration());
```
5. Run dotnet build to verifty if it passes build.
```
EchoBot-9\runtime\azurewebapp>dotnet build
```
6. Go to azurewebapp/schemas folder, run update.cmd, the sdk.schema will be updated.
```
EchoBot-9\runtime\azurewebapp\schemas>update.cmd
```
7. Copy the sdk.schema to the bot/schemas
8. Go back to composer, reopen the bot and you can add a custom action.  

## Task Item

Create your bot based on custom actions and try out in emulator. 

fixes #2824 
fixes #2896 
<!---
Please include a link to the related issue. [Ex. `Closes #<issue #>`](https://help.github.com/en/articles/closing-issues-using-keywords)
---->

## Screenshots
![image](https://user-images.githubusercontent.com/32191031/81674612-0f33d000-9480-11ea-83cd-e745ce3cf417.png)

![image](https://user-images.githubusercontent.com/32191031/81674694-2e326200-9480-11ea-8779-7d373f6c5478.png)

![image](https://user-images.githubusercontent.com/32191031/81674736-41453200-9480-11ea-8c72-88761209f73a.png)

![image](https://user-images.githubusercontent.com/32191031/81674765-4d30f400-9480-11ea-8905-3e6faed5ccad.png)

![image](https://user-images.githubusercontent.com/32191031/81675341-435bc080-9481-11ea-9e07-6f73ec766148.png)

![image](https://user-images.githubusercontent.com/32191031/81675362-4951a180-9481-11ea-857c-b7d77cd4734b.png)






<!---
Please include screenshots or gifs if your PR include UX changes.
--->

boydc2014 commented on Composer/plugins/samples/assets/projects/CustomAction/customaction.dialog:
```diff
@@ -0,0 +1,100 @@
+{
```
I thought since we modified the built-in CSharp template (code + schema), we don't actually need this sample,  every sample can be used as customized action sample.

boydc2014 replied:
see https://github.com/microsoft/BotFramework-Composer/issues/2824#issuecomment-621603697

boydc2014 commented on Composer/plugins/samples/assets/projects/CustomAction/schemas/sdk.schema:
```diff
@@ -0,0 +1,8901 @@
+{
```
this should be an update to the default sdk.schema, instead of creating a new one.

boydc2014 commented on Composer/plugins/samples/assets/projects/CustomAction/schemas/MultiplyDialog.schema:
```diff
@@ -0,0 +1,25 @@
+﻿{
```
This is unnecessary, and doing the schema here and doing the code in another place actually make the CSharp Bot template not "self-contained".

luhan2017 commented:
If we want to keep the custom schema "custom", we need keep two versions of schema in composer:
1. composer/packages/server/schemas/sdk.schema, this is our default schema and should not include custom schema. 
2. runtime/azurewebapp/schemas/sdk.schema. this is default+custom schema.

which means, every time we update the sdk package+schema, we need to keep in mind to do twice of the schema merge for the above two, which doesn't look good for me.

So my hacky way is to put custom schema in the sample folder and when ejection you can customize you sdk.schema by rerun the dialog:merge.

I think both way are not perfect, how about we just explain how to use it in documents. I am not sure providing a single click experience on custom schema is a good direction which will make user confused.  ("I don't need to do anything to enable custom schema"?)

@srinaath is working on a readme to explain how to enable this. we can decide whether that is enough for user. if adding this sample just introduces more confusion, let's close this. 

@cwhitten @benbrown  thoughts?

hibrenda commented:
I prefer to let user know how to merge the sdk.schema instead of just give them the customized sdk.schema as if user really use this feature later, he need to know how to do the schema merge.
So until we have the UX trade off to implement this step later, we just provide the readme file(manual step) to user. 

For custom schema, there are 2 steps. 
1. custom runtime (self-contained now)
- The customized runtime can be set up through the ejection 
- User can learn from this sample for how to write a customized code and also how to register the customized code into startup.cs . User can easily extend it based on this sample code.

2. customized schema (manual for R9)
- The customized schema (sample) is located in the sample folder  
- User need to follow readme to install bf-dialog tool and manually execute "bf dialog:merge" to get the updated sdk.schema file.
* Just like we did the manual step for azure provision, if user don't follow the step, they can not make it.

srinaath commented:
@luhan2017  I used this PR to actually build my README file for adding custom actions. I think there is a lot of value in including this sample and I'm mentioning about this sample in the README too. Granted the PR does a little magic of showing the merged schema directly. But the README coupled with this sample should give a user a good idea to add custom actions.

srinaath commented on runtime/dotnet/azurewebapp/Startup.cs:
```diff
@@ -168,6 +169,9 @@ public void ConfigureServices(IServiceCollection services)
             var rootDialog = GetRootDialog(botDir);
 
+            // register customized types, please note this is only needed for customized actions. 
+            resourceExplorer.RegisterType<MultiplyDialog>("MultiplyDialog");
```
@luhan2017 so currently the default runtime that is ejected out is going to include this registration for all samples correct? Do you think if we could include this registration only if they eject out CustomActions sample?

benbrown replied:
yes we should include it for everyone so they have a clear example of how it works.

carlosscastro commented:
@luhan2017 @benbrown should this be added to the functions runtime as well? If yes, we could place the dialog in the core project (/runtime/core) and then jsut reference it from the web app startup.cs and from the functions startup.cs.

benbrown commented:
IMO - yes!

luhan2017 commented:
To who evolved in the PR, @benbrown @benbrown @hibrenda @cwhitten @srinaath 

I have removed an action sample and added a custom action project in the runtime folder. which user need to refer to it explicitly to enable it. See the details at the beginning. 

@carlosscastro , if user want to enable it for azure function, he/she will need to add a project reference for azure function and add CustomActionComponentRegistration in Azurefunction/startup.cs. which I didn't explicitly add right now.

srinaath commented:
Awesome @luhan2017 .. Adding my README commits to this PR too. Once added I think we can request a rereview and merge it

srinaath commented:
README updated and update.sh script to run bfdialog tool on Mac/Win added. PR ready for rereview
